sbi/inference/posteriors/direct_posterior.py

```diff
@@ -171,24 +168,27 @@ def log_prob(
             support of the prior, -∞ (corresponding to 0 probability) outside.
         """
         x = self._x_else_default_x(x)
-        condition_shape = self.posterior_estimator._condition_shape
-        try:
-            x = x.reshape(*condition_shape)
-        except RuntimeError as err:
-            raise ValueError(
-                f"Expected a single `x` which should broadcastable to shape \
-                  {condition_shape}, but got {x.shape}. For batched eval \
-                  see issue #990"
-            ) from err
 
-        # TODO Train exited here, entered after sampling?
-        self.posterior_estimator.eval()
+        # [1:] to remove batch dimension for `reshape_to_iid_batch_event`.
+        x_shape = self._x_shape[1:] if self._x_shape is not None else x.shape[1:]
 
         theta = ensure_theta_batched(torch.as_tensor(theta))
+        theta_de = reshape_to_iid_batch_event(
+            theta, theta.shape[1:], leading_is_iid=True
+        )
+        x_de = reshape_to_batch_event(x, x_shape)
+
+        # TODO Train exited here, entered after sampling?
```

Comment:
Remove? It is fine to ensure eval here, although it should already be set to this mode after training.